File: 155.最小栈.py
#
# @lc app=leetcode.cn id=155 lang=python3
#
# [155] 最小栈
#

# @lc code=start
from typing import List
import math


class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self._arr: List[int] = []
        self._minArr: List[int] = [math.inf]

    def push(self, val: int) -> None:
        self._arr.append(val)
        self._minArr.append(min(self.getMin(), val))

    def pop(self) -> None:
        self._arr.pop()
        self._minArr.pop()

    def top(self) -> int:
        return self._arr[-1]

    def getMin(self) -> int:
        return self._minArr[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @lc code=end

File: 155.最小栈.ts
/*
 * @lc app=leetcode.cn id=155 lang=typescript
 *
 * [155] 最小栈
 */

// @lc code=start
class MinStack {
    private arr: number[] = [];
    private minArr: number[] = [Infinity];

    constructor() {}

    push(val: number): void {
        this.arr.push(val);
        this.minArr.push(Math.min(this.getMin(), val));
    }

    pop(): void {
        this.arr.pop();
        this.minArr.pop();
    }

    top(): number {
        return this.arr[this.arr.length - 1];
    }

    getMin(): number {
        return this.minArr[this.minArr.length - 1];
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
// @lc code=end
